modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +182,102 @@ private void ensureOpen() {
     }
 
     private void reopenStreamOrFail(IOException e) throws IOException {
-        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
-            logger.warn(() -> format("""
-                failed reading [%s/%s] at offset [%s]""", blobStore.bucket(), blobKey, start + currentOffset), e);
-            throw e;
+        if (shouldRetry() == false) {
+            final var finalException = addSuppressedExceptions(e);
+            logForRetry(Level.WARN, "reading", finalException);
+            throw finalException;
         }
 
-        final int maxAttempts = blobStore.getMaxRetries() + 1;
+        attempt += 1;
+        // Log at info level every ~5 minutes
+        logForRetry(attempt % 30 == 0 ? Level.INFO : Level.DEBUG, "reading", e);
+        if (failures.size() < MAX_SUPPRESSED_EXCEPTIONS) {
+            failures.add(e);
+        }
```

Comment:
nit All this code up to this point also appears similarly above below line 105. I wonder if maybe this should be refactored into another function like `checkCanRetry(action, e)` or something like that.

Reply:
I extracted a new method `recordAndLogForRetry` for these and the increment of `attempt`.